### ISSUE TITLE
ObservableSubscriberProxy doesn't allow contravariant Observers

### ISSUE THREAD

**Issue by tadfisher** [uber/AutoDispose#243]
**Library version**: 1.0.0-RC1

The signature of `ObservableSource#subscribe` is as follows:

```java
void subscribe(@NonNull Observer<? super T> observer);
```

However, `ObservableSubscriberProxy` doesn't allow contravariant Observer types:

```java
void subscribe(Observer<T> observer);
```

This is mainly an issue when using sealed classes in Kotlin, as implementing `ObservableSource<T>`  where `T` is abstract forces the subscription parameter type to `Observer<in T>`, which the proxy subscribe method doesn't allow.

TLDR; This works:

```kotlin
override fun subscribe(observer: Observer<in AbstractType>) {
    Observable.just(ConcreteType())
        .subscribe(observer)
}
```

But this doesn't:

```kotlin
override fun subscribe(observer: Observer<in AbstractType>) {
    Observable.just(ConcreteType())
        .autoDisposable(scope())
        .subscribe(observer)
}
```
